core/src/autogluon/core/utils/feature_selection.py

```diff
@@ -0,0 +1,497 @@
+            return acceptance_candidates
+
+        # Try removing features with lowest importance first
+        removal_candidate_rows = below_threshold_rows[:n_remove]
+        removal_candidates = tuple(removal_candidate_rows.index)
+        if removal_candidates not in self.attempted_removals:
+            acceptance_candidates = importance_df[~importance_df.index.isin(removal_candidates)].index.tolist()
+            self.attempted_removals.add(removal_candidates)
+            return acceptance_candidates
+
+        sample_weights = [1/i for i in range(1, len(below_threshold_rows)+1)] if weighted else None
+        while time_budget - (time.time() - time_start) > 0:
+            random_state = self.rng.integers(low=0, high=1e5)
+            removal_candidate_rows = below_threshold_rows.sample(n=n_remove, random_state=random_state, replace=False, weights=sample_weights)
+            removal_candidates = tuple(removal_candidate_rows.index)
+            if removal_candidates not in self.attempted_removals:
+                acceptance_candidates = importance_df[~importance_df.index.isin(removal_candidates)].index.tolist()
+                self.attempted_removals.add(removal_candidates)
+                return acceptance_candidates
+        return importance_df.index.tolist()
+
+    def compute_expected_fi_time_single(self, X_fi: pd.DataFrame, model_predict_time: float, n_subsample: int, n_total_sample: int) -> float:
+        """
+        Return the expected time to compute permutation feature importance for a single feature on n_total_sample datapoints.
+        """
+        n_subsample = min(n_subsample, len(X_fi))
+        n_shuffle = min(np.ceil(n_total_sample / n_subsample).astype(int), self.max_n_shuffle)
+        return self.safety_time_multiplier * model_predict_time * (n_subsample / len(X_fi)) * n_shuffle
+
+    def compute_time_budget_fi(self, X_fi: pd.DataFrame, n_subsample: int, **kwargs):
+        """
+        Return the time that a single feature importance computation round can take up to. Currently the minimum of 5 minutes
+        and the time it takes to fully evaluated minimum of 50 features or the number of features in the dataset.
+        """
+        min_fi_samples = kwargs.get('min_fi_samples', 10000)
+        max_fi_samples = kwargs.get('max_fi_samples', 100000)
+        n_total_samples = max(min_fi_samples, min(max_fi_samples, len(X_fi)))
+        fi_time_single = self.compute_expected_fi_time_single(X_fi, self.model_predict_time, n_subsample, n_total_samples)
+        return min(fi_time_single * min(len(X_fi.columns), 50), self.max_time_budget_fi)
+
+    def fit_score_model(self, X: pd.DataFrame, y: pd.Series, X_val: pd.DataFrame, y_val: pd.Series,
+                        features: List[str], model_name: str, **kwargs) -> Tuple[AbstractModel, float, float]:
+        """
+        Fits and scores a model. Updates self.time_limit. Returns the fitted model, its score, and time elapsed.
```

Comment:
description could clarify how this relates to feature pruning, eg.
"Fits and scores a model over the remaining feature subset in a particular feature-pruning round"